staging/kos/pkg/controllers/connectionagent/connection_agent.go

```diff
@@ -0,0 +1,1114 @@
+	// deleted and replaced the status.hostIP field of the newer attachment is
+	// set to "", and the connection agent has to write back the correct value.
+	// Also, the only fields affecting local network interfaces handling that
+	// can be seen changing by this function are status.ipv4 and
+	// status.addressVNI, so enqueue if they changed.
+	if oldAtt.UID != att.UID || oldAtt.Status.IPv4 != att.Status.IPv4 || oldAtt.Status.AddressVNI != att.Status.AddressVNI {
+		ca.queue.Add(parse.AttNSN(att))
+	}
+}
+
+func (ca *ConnectionAgent) onLocalAttDelete(obj interface{}) {
+	att := parse.Peel(obj).(*netv1a1.NetworkAttachment)
+	klog.V(5).Infof("Local NetworkAttachments cache: notified of removal of %#+v", att)
+
+	attNSN := parse.AttNSN(att)
+	ca.updateL1VNStateForLocalAtt(attNSN, false)
+	ca.queue.Add(attNSN)
+}
+
+func (ca *ConnectionAgent) updateL1VNStateForLocalAtt(att k8stypes.NamespacedName, attExists bool) {
+	ca.l1VirtNetsState.Lock()
+	defer ca.l1VirtNetsState.Unlock()
+
+	attListersIDs := ca.l1VirtNetsState.attToListersIDs[att]
+
+	if attExists {
+		if attListersIDs == nil {
+			attListersIDs = make(map[uint32]struct{}, 1)
+			ca.l1VirtNetsState.attToListersIDs[att] = attListersIDs
+		}
+		attListersIDs[localAttsListerID] = struct{}{}
+		return
+	}
+
+	delete(attListersIDs, localAttsListerID)
+	if len(attListersIDs) == 0 {
+		delete(ca.l1VirtNetsState.attToListersIDs, att)
+	}
+}
+
+func (ca *ConnectionAgent) syncPreExistingNetworkInterfaces() error {
+	// Start all the remote attachments informers because to choose whether to
+	// keep a pre-existing remote network interface we need to look for a remote
+	// network attachment that can own it in the informer cache for the VNI of
+	// the network interface.
+	err := ca.startRemoteAttsInformers()
+	if err != nil {
+		return fmt.Errorf("failed to sync pre-existing network interfaces: %s", err.Error())
+	}
+
+	ifcs, err := ca.listPreExistingNetworkInterfaces()
+	if err != nil {
+		return fmt.Errorf("failed to sync pre-existing network interfaces: %s", err.Error())
+	}
+
+	for _, ifc := range ifcs {
+		err = ca.syncPreExistingNetworkInterface(ifc)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (ca *ConnectionAgent) startRemoteAttsInformers() error {
+	localAtts, err := ca.localAttsLister.List(k8slabels.Everything())
+	if err != nil {
+		return fmt.Errorf("failed to list local network attachments: %s", err.Error())
+	}
+
+	for _, att := range localAtts {
+		// Adding `att` to the L2 virtual network state entails starting the
+		// remote attachments informer for `att`'s vni. addLocalAttToL2VNState
+		// returns an error (and does not perform the addition) if `att`'s
+		// namespace differs from the one recorded in the L2VNState for `att`'s
+		// vni. The namespace recorded in such L2VNState is the namespace of the
+		// first attachment added to it. If an error is returned, we're during a
+		// transient where two or more NetworkAttachments with same VNI but
```

Comment:
Could we preclude this transient by making subnet validation contingent on absence of conflicting NetworkAttachment objects?  If so, let's do that in a subsequent PR.

Reply:
Let me make sure I get this right. Your thinking about something along the lines of
"Subnet X gets validated if and only if it does not conflict with other subnets and if and only if there are no NetworkAttachments with X's VNI in namespaces other than X's (on the basis that such NetworkAttachments belonged to a previously existing subnet with X's VNI but different namespace)". I understand the temptation, but I'd rather not do that. It would require heavy re-design of the subnet validator, which would have to be instrumented to deal with NetworkAttachments as well. Also, if a subnet X cannot be validated because of a NetworkAttachment Y, when Y is deleted X should be revalidated; but Y's fields offer nothing to retrieve the subnets whose validation might be pending because of Y (TBH `status.addressVNI` carries that information, but it gets cleared after Y's subnet is deleted), and re-processing all subnets is ugly and inefficient. I'm not sure the advantage of not having to handle the transient in the connection agent is worth such a complication of the subnet validator.

Reply:
I agree.